src/typechecker.erl

```diff
@@ -3401,18 +3408,32 @@ check_clauses(Env, ArgsTy, ResTy, Clauses, Caps) ->
     check_exhaustiveness(Env, ArgsTy, Clauses, RefinedArgsTy, VarBindsList, Css).
 
 check_exhaustiveness(Env, ArgsTy, Clauses, RefinedArgsTy, VarBindsList, Css) ->
-    case {Env#env.exhaust,
-          ArgsTy =/= any andalso lists:all(fun (Ty) -> refinable(Ty, Env) end, ArgsTy),
-          lists:all(fun no_guards/1, Clauses),
-          is_list(RefinedArgsTy) andalso lists:any(fun (T) -> T =/= type(none) end, RefinedArgsTy)} of
-        {true, true, true, true} ->
+    case is_non_exhaustive(Env#env.exhaust,
+
+                           ArgsTy =/= any
+                           andalso lists:all(fun (Ty) -> refinable(Ty, Env) end, ArgsTy),
+
+                           lists:all(fun no_guards/1, Clauses),
+
+                           is_list(RefinedArgsTy)
+                           andalso lists:any(fun (T) -> T =/= type(none) end, RefinedArgsTy))
+    of
+        true ->
```

Comment:
TBH, I don't think this change improves on clarity.

Adding line breaks (and comments?) can be useful, but moving the logic to a function below mostly hides what's happening from the view.

Another style I'd be fine with is

```Erlang
if
    Env#env.exhaus andalso
    ... andalso
    ... ->
        %% do exhaustiveness-checking
    true ->
        %% no exhaustiveness-checking
```

Reply:
To be clear, I want the intermediate result printed as a separate trace message, to understand which path is taken and why:

```erlang
{trace,<0.9.0>,call,{typechecker,is_non_exhaustive,[true,true,true,true]}}

{trace,<0.9.0>,return_from,{typechecker,is_non_exhaustive,4},true}
```

That's a tiny snippet, but consider:

```
{trace,<0.9.0>,call,
    {typechecker,type_check_forms,
        [[{attribute,
              {1,1},
              file,
              {"test/should_fail/nonempty_list_match_in_head_nonexhaustive.erl",
               1}},
          {attribute,{1,2},module,nonempty_list_match_in_head_nonexhaustive},
          {attribute,{3,2},export,[{f,1}]},
          {attribute,
              {12,2},
              type,
              {t,{type,
                     {12,14},
                     union,
                     [{type,{12,14},tuple,[]},
                      {type,{12,19},list,[{user_type,{12,20},t,[]}]}]},
                 []}},
          {attribute,
              {14,2},
              spec,
              {{f,1},
               [{type,
                    {14,8},
                    'fun',
                    [{type,{14,8},product,[{user_type,{14,9},t,[]}]},
                     {atom,{14,17},ok}]}]}},
          {function,
              {15,1},
              f,1,
              [{clause,
                   {15,1},
                   [{cons,{15,3},{var,{15,4},'_'},{var,{15,6},'_'}}],
                   [],
                   [{atom,{15,13},ok}]}]},
          {eof,{16,1}}],
         []]}}

...

{trace,<0.9.0>,return_from,
       {typechecker,mta,2},
       {type,0,list,[{user_type,0,t,[]}]}}

{trace,<0.9.0>,return_from,
       {typechecker,stop_refinable_recursion,3},
       {proceed,{set,3,16,16,8,80,48,
                     {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
                     {{[],[],[],[],[],[],
                       [{nonempty_list_match_in_head_nonexhaustive,t,0}],
                       [],[],[],[],[],[],
                       [{type,0,union,
                              [{type,0,tuple,[]},
                               {type,0,list,[{user_type,0,t,[]}]}]}],
                       [{type,0,list,[{user_type,0,t,[]}]}],
                       []}}}}}

{trace,<0.9.0>,call,
    {typechecker,refinable,
        [{user_type,0,t,[]},
         {env,
             #{{f,1} =>
                   [{type,0,bounded_fun,
                        [{type,0,'fun',
                             [{type,0,product,[{user_type,0,t,[]}]},
                              {atom,0,ok}]},
                         []]}]},
             #{},#{},
             #{module => nonempty_list_match_in_head_nonexhaustive,
               records => #{},
               types =>
                   #{{t,0} =>
                         {[],
                          {type,0,union,
                              [{type,0,tuple,[]},
                               {type,0,list,[{user_type,0,t,[]}]}]}}}},
             false,false,true,30},
         {set,3,16,16,8,80,48,
             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
             {{[],[],[],[],[],[],
               [{nonempty_list_match_in_head_nonexhaustive,t,0}],
               [],[],[],[],[],[],
               [{type,0,union,
                    [{type,0,tuple,[]},{type,0,list,[{user_type,0,t,[]}]}]}],
               [{type,0,list,[{user_type,0,t,[]}]}],
               []}}}]}}

{trace,<0.9.0>,call,
    {typechecker,stop_refinable_recursion,
        [{user_type,0,t,[]},
         {env,
             #{{f,1} =>
                   [{type,0,bounded_fun,
                        [{type,0,'fun',
                             [{type,0,product,[{user_type,0,t,[]}]},
                              {atom,0,ok}]},
                         []]}]},
             #{},#{},
             #{module => nonempty_list_match_in_head_nonexhaustive,
               records => #{},
               types =>
                   #{{t,0} =>
                         {[],
                          {type,0,union,
                              [{type,0,tuple,[]},
                               {type,0,list,[{user_type,0,t,[]}]}]}}}},
             false,false,true,30},
         {set,3,16,16,8,80,48,
             {[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},
             {{[],[],[],[],[],[],
               [{nonempty_list_match_in_head_nonexhaustive,t,0}],
               [],[],[],[],[],[],
               [{type,0,union,
                    [{type,0,tuple,[]},{type,0,list,[{user_type,0,t,[]}]}]}],
               [{type,0,list,[{user_type,0,t,[]}]}],
               []}}}]}}

{trace,<0.9.0>,call,
    {typechecker,mta,
        [{user_type,0,t,[]},
         {env,
             #{{f,1} =>
                   [{type,0,bounded_fun,
                        [{type,0,'fun',
                             [{type,0,product,[{user_type,0,t,[]}]},
                              {atom,0,ok}]},
                         []]}]},
             #{},#{},
             #{module => nonempty_list_match_in_head_nonexhaustive,
               records => #{},
               types =>
                   #{{t,0} =>
                         {[],
                          {type,0,union,
                              [{type,0,tuple,[]},
                               {type,0,list,[{user_type,0,t,[]}]}]}}}},
             false,false,true,30}]}}

{trace,<0.9.0>,return_from,
       {typechecker,mta,2},
       {nonempty_list_match_in_head_nonexhaustive,t,0}}

{trace,<0.9.0>,return_from,{typechecker,stop_refinable_recursion,3},stop}

{trace,<0.9.0>,return_from,{typechecker,refinable,3},true}

{trace,<0.9.0>,return_from,{typechecker,refinable,3},true}

{trace,<0.9.0>,return_from,{typechecker,'-refinable/3-fun-0-',3},true}

{trace,<0.9.0>,return_from,{typechecker,refinable,3},true}

{trace,<0.9.0>,return_from,{typechecker,refinable,3},true}

{trace,<0.9.0>,return_from,{typechecker,refinable,2},true}

{trace,<0.9.0>,return_from,
       {typechecker,'-check_exhaustiveness/6-fun-2-',2},
       true}

{trace,<0.9.0>,call,
       {typechecker,'-fun.no_guards/1-',
                    [{clause,{15,1},
                             [{cons,{15,3},{var,{15,4},'_'},{var,{15,6},'_'}}],
                             [],
                             [{atom,{15,13},ok}]}]}}

{trace,<0.9.0>,call,
       {typechecker,no_guards,
                    [{clause,{15,1},
                             [{cons,{15,3},{var,{15,4},'_'},{var,{15,6},'_'}}],
                             [],
                             [{atom,{15,13},ok}]}]}}

{trace,<0.9.0>,return_from,{typechecker,no_guards,1},true}

{trace,<0.9.0>,return_from,{typechecker,'-fun.no_guards/1-',1},true}

{trace,<0.9.0>,call,
       {typechecker,'-check_exhaustiveness/6-fun-0-',[{user_type,0,t,[]}]}}

{trace,<0.9.0>,call,{typechecker,type,[none]}}

{trace,<0.9.0>,call,{typechecker,type,[none,[]]}}

{trace,<0.9.0>,return_from,{typechecker,type,2},{type,0,none,[]}}

{trace,<0.9.0>,return_from,{typechecker,type,1},{type,0,none,[]}}

{trace,<0.9.0>,return_from,
       {typechecker,'-check_exhaustiveness/6-fun-0-',1},
       true}

{trace,<0.9.0>,call,{typechecker,is_non_exhaustive,[true,true,true,true]}}

{trace,<0.9.0>,return_from,{typechecker,is_non_exhaustive,4},true}

{trace,<0.9.0>,exception_from,
       {typechecker,check_exhaustiveness,6},
       {throw,{nonexhaustive,{15,1},[{tuple,0,[]}]}}}

{trace,<0.9.0>,exception_from,
       {typechecker,check_clauses,5},
       {throw,{nonexhaustive,{15,1},[{tuple,0,[]}]}}}

{trace,<0.9.0>,exception_from,
       {typechecker,check_clauses_fun,3},
       {throw,{nonexhaustive,{15,1},[{tuple,0,[]}]}}}

{trace,<0.9.0>,exception_from,
       {typechecker,type_check_function,2},
       {throw,{nonexhaustive,{15,1},[{tuple,0,[]}]}}}

{trace,<0.9.0>,return_from,
       {typechecker,'-type_check_forms/2-fun-0-',5},
       [{nonexhaustive,{15,1},[{tuple,0,[]}]}]}

{trace,<0.9.0>,return_from,
       {typechecker,type_check_forms,2},
       [{nonexhaustive,{15,1},[{tuple,0,[]}]}]}

test/should_fail/nonempty_list_match_in_head_nonexhaustive.erl: Nonexhaustive patterns on line 15 at column 1
Example values which are not covered:
	{}
```

This trace, from calling `typechecker,type_check_forms` to returning, is ~22k lines long. And that's running the typechecker on a minimised test example.

Reply:
To be even more clear, I don't care that much about the indentation in between the new function arguments. The important part, of course, is extracting the compound condition to a new function, so it's traceable.

Reply:
I get it. :-)

Reply:
In that case, I'd prefer functions with more meaningful names and parameters. Adding functions should also improve readability of the code, not only tracing.

Idea:

```Erlang
case all_refinable(ArgsTy, Env) andalso
     no_clause_has_guards(Clauses) andalso
     some_type_is_not_none(RefinedArgsTy) of
    true when Env#env.exhaust ->
        ...
```

where the implementations are something like

```Erlang
all_refineable(any, _Env) -> false;
all_refineable(Types, Env) -> lists:all(fun (Ty) -> refinable(Ty, Env) end, Types).

no_clause_has_guards(Clauses) ->
    lists:all(fun no_guards/1, Clauses).

some_type_is_not_none(Types) ->
    lists:any(fun (T) -> T =/= type(none) end, Types).
```

It's still good for tracing, isn't it?

Btw, documenting the what this function does and what the parameters are would be very useful:

```Erlang
%% Checks if clauses completely covers the types of the arguments and wtf about argstys refinedargsty(?)
check_exhaustiveness(Env, ArgsTy, Clauses, RefinedArgsTy, VarBindsList, Css)
```

Reply:
Good idea, I've refactored the code and added some docs 👍